src/webview/__tests__/generateInboundEventEditSequence-test.js

```diff
@@ -590,4 +707,147 @@ describe('getEditSequence correct for interesting changes', () => {
+          },
+        ],
+      };
+      const msgWithChoice = {
+        ...baseMessage,
+        submessages: [
+          ...msgWithPoll.submessages,
+          {
+            ...baseSubmessage,
+            content: '{"type":"new_option","idx":1,"option":"Choice A"}',
+            id: 2,
+          },
+        ],
+      };
+
+      test('choice added', () => {
+        check({ messages: [msgWithPoll] }, { messages: [msgWithChoice] });
+      });
+
+      const msgWithVote = {
+        ...baseMessage,
+        submessages: [
+          ...msgWithChoice.submessages,
+          {
+            ...baseSubmessage,
+            content: `{"type":"vote","key":"${baseMessage.sender_id},1","vote":1}`,
+            id: 3,
+          },
+        ],
+      };
+
+      test('vote added', () => {
+        check({ messages: [msgWithChoice] }, { messages: [msgWithVote] });
+      });
+    });
+
+    test('star a message', () => {
+      const message = eg.streamMessage();
+      check(
+        {
+          messages: [message],
+          backgroundData: {
+            ...eg.backgroundData,
+            flags: { ...eg.backgroundData.flags, starred: {} },
+          },
+        },
+        {
+          messages: [message],
+          backgroundData: {
+            ...eg.backgroundData,
+            flags: { ...eg.backgroundData.flags, starred: { [message.id]: true } },
+          },
+        },
+      );
+    });
+
+    test('unstar a message', () => {
+      const message = eg.streamMessage();
+      check(
+        {
+          messages: [message],
+          backgroundData: {
+            ...eg.backgroundData,
+            flags: { ...eg.backgroundData.flags, starred: { [message.id]: true } },
+          },
+        },
+        {
+          messages: [message],
+          backgroundData: {
+            ...eg.backgroundData,
+            flags: { ...eg.backgroundData.flags, starred: {} },
+          },
+        },
+      );
+    });
+
+    // TODO(#5208): We haven't settled how we want to track name/avatar
+    test.todo("sender's name/avatar changed");
+
+    test('mute a sender', () => {
+      const message = eg.streamMessage();
+      check(
+        {
+          messages: [message],
+          backgroundData: { ...eg.backgroundData, mutedUsers: Immutable.Map() },
+        },
+        {
+          messages: [message],
+          backgroundData: {
+            ...eg.backgroundData,
+            mutedUsers: Immutable.Map([[message.sender_id, 1644366787]]),
+          },
+        },
+      );
+    });
+
+    test('unmute a sender', () => {
+      const message = eg.streamMessage();
+      check(
+        {
+          messages: [message],
+          backgroundData: {
+            ...eg.backgroundData,
+            mutedUsers: Immutable.Map([[message.sender_id, 1644366787]]),
+          },
+        },
+        {
+          messages: [message],
+          backgroundData: { ...eg.backgroundData, mutedUsers: Immutable.Map() },
+        },
+      );
```

Comment:
Seeing the test cases like this one, here's one thought that occurs to me. This idea isn't really a request for this PR -- I'd actually rather go ahead and merge a version of this PR close to how it currently is and then have this refactor come in the next PR.

I think that rather than have `check` take an old and a new BackgroundData value, it'd be better to have it take an old and a new PerAccountState, from which it would compute the BackgroundData.

One reason is that that would feel a bit more end-to-end -- I'd worry a bit less about a gap between what these tests are testing, and what we actually do as events come in and update our state.

In particular, we could factor out the selector MessageList uses for computing the BackgroundData, and export it only for tests, and call it from here. (I guess it would also need to take a GlobalSettings and a Debug, as well as a PerAccountState. But `check` could still take just the PerAccountState, and supply the other two because they're boring here.) But even if `check` were to basically copy-paste that chunk of `MessageList.js`, which looks like this:
```js
    const backgroundData: BackgroundData = {
      alertWords: state.alertWords,
      allImageEmojiById: getAllImageEmojiById(state),
      auth: getAuth(state),
```
then I feel like that'd be an improvement; it'd mostly be pretty hard for a copy of that to become out of sync and wrong without the type-checker telling us so.

Then a further thing that that would enable is that because it'd make this test data closer to the form in which we actually construct and update the data, it'd let us construct it using that same code -- that is, we could construct it using the reducers, rather than with object literals and so on. (That's a direction you've seen me move in for other tests in some recent PRs.)

That would further enhance the end-to-end-ness of these tests. It'd also make it so that they don't have to change so much whenever we rearrange things in the state.

The ideal version might even be that each of these test cases sets up some state, then sets up the new state by hitting that not with one of our Redux actions but with a server event. That's the one really stable API that this all needs to work to, after all -- everything inward of that, we can freely refactor on a whim. But doing it with reducers is basically directly on the path to that, and doing it with state objects (however constructed) is in turn on the path to that, so for the next PR I'd be happy to see just switching from BackgroundData to state objects as the arguments passed to `check`.

Reply:
Cool! I'm excited to go in that direction for the tests in a separate PR.